rls/rls-cargo-build-integration-notes.md

```diff
@@ -0,0 +1,105 @@
+  - triggerd by the RLS (source files changed)
+#### From scratch
+The build plan has possibly changed or is insufficient - we need to regenerate it using the build system.
+  - triggered by the user (e.g. build manifest or relevant package configuration was changed)
+
+## Cargo
+Cargo is capable of generating a build plan for `cargo build` (with `-Z unstable-options --build-plan`), which is a dependency graph between commands (`rustc` and calling build scripts, together with args, env, cwd), where each command is also tagged with a package name, version, crate target kind, Target/Host architecture differentiation and a list of resulting crate artifacts.
+
+Generating the build plan and its format are currently unstable and not yet used to integrate with other build systems or RLS.
+
+Note: When emitting build plan for `cargo check` is supported, the RLS could possibly drop a dependency the Cargo library.
+
+### What Cargo does (`cargo {build,check}`)
+An arbitrary breakdown that we think might be useful to identify separate, replaceable steps:
+1. Configuration (incl. package overrides etc.)
+2. Version resolution
+3. Package dependency graph construction (Cargo.lock)
+4. Action graph construction
+    1. Feature resolution - may change package deps based on features selected
+    2. Lowering dependencies to crate target level (proc macros, bins, libs)
+    3. Spliting nodes further into actions (e.g. compile build script, run build script, compile lib)
+    4. Differentiate/prune actions based on target dirtiness (fingerprints)
+5. Fetching packages (locate on disk, fetch from network)
+6. Build plan construction - flattened view into action graph mapped into actual system commands
+7. Execute the commands!
+
+# Possible plan
+## Split Cargo into 3 parts
+Having the above sequence of steps for `cargo build`, we can group these into:
+1. Front (`cargo build` -> action graph) (Step 1-5)
+2. Build orchestration (action graph -> build plan) (Step 6)
+3. Execution (invoking the build plan commands) (Step 7)
+
+With this we can mix and match:
+1. RLS should always execute the build (needs to invoke `rustc` itself to perform analysis)
+2. External build system can provide its set of packages and dependencies between them and Cargo can orchestrate that into a build plan (understandable by the RLS)
+3. Cargo can execute its front, fetch packages from crates.io, form an action graph, which can be further reused and translated into a build plan by the external build system
+4. etc.
+
+## Concrete goals
+### RLS <> other build systems
+* Establish a preliminary build plan format (probably use/adapt current Cargo `--build-plan`)
+* RLS needs to understand the build plan and be able to execute it in a sandbox (think `target/rls` dir), while consuming the save-analysis and diagnostics
+### RLS <> Cargo
+* Implement `cargo check ... --build-plan` in Cargo
```

Comment:
Probably want to add in here the RLS asking Cargo to run the build plan, but having the RLS execute the steps